lib/core/src/Cardano/Wallet.hs

```diff
@@ -331,33 +374,58 @@ mkWalletLayer db nw tl = WalletLayer
                                      Keystore
     ---------------------------------------------------------------------------}
 
-    , attachPrivateKey = \wid (xprv, pwd) -> do
-        hpwd <- liftIO $ encryptPassphrase pwd
-        DB.putPrivateKey db (PrimaryKey wid) (xprv, hpwd)
+    , attachPrivateKey = _attachPrivateKey
     }
   where
     _readWallet
         :: WalletId
         -> ExceptT ErrNoSuchWallet IO (Wallet s t, WalletMetadata)
-    _readWallet wid = maybeToExceptT (ErrNoSuchWallet wid) $ do
-        cp <- MaybeT $ DB.readCheckpoint db (PrimaryKey wid)
-        meta <- MaybeT $ DB.readWalletMeta db (PrimaryKey wid)
-        return (cp, meta)
+    _readWallet wid = (,)
+        <$> _readWalletCheckpoint wid
+        <*> _readWalletMeta wid
+
+    _readWalletMeta
+        :: WalletId
+        -> ExceptT ErrNoSuchWallet IO WalletMetadata
+    _readWalletMeta wid = maybeToExceptT (ErrNoSuchWallet wid) $
+        MaybeT $ DB.readWalletMeta db (PrimaryKey wid)
+
+    _readWalletCheckpoint
+        :: WalletId
+        -> ExceptT ErrNoSuchWallet IO (Wallet s t)
+    _readWalletCheckpoint wid = maybeToExceptT (ErrNoSuchWallet wid) $ do
+        MaybeT $ DB.readCheckpoint db (PrimaryKey wid)
+
+    _attachPrivateKey
+        :: WalletId
+        -> (Key 'RootK XPrv, Passphrase "encryption")
+        -> ExceptT ErrNoSuchWallet IO ()
+    _attachPrivateKey wid (xprv, pwd) = do
+        hpwd <- liftIO $ encryptPassphrase pwd
+        DB.putPrivateKey db (PrimaryKey wid) (xprv, hpwd)
```

Comment:
could this fail due to the race condition?

Reply:
If there's no wallet? Yes, it would fail with `ErrNoSuchWallet`. Could happen if the wallet is deleted before the key gets added. That's `okay` and already handled.

Reply:
`MVar` implementation shouldn't. I hope sql version won't as well